Rhel7 support

felixb opened the pull request:
Add support for RHEL7/ CentOS7 with systemd.

JayH5 commented:
Thanks for the pull request. :smile: 

It is quite a big PR so you'll have to give us a few days to review it. We don't run RHEL or CentOS so we'll have to take your word that you've tested it and it works on those OSes. Otherwise, it looks good - thanks for adding some tests.

JayH5 commented on manifests/config.pp:
```diff
@@ -78,11 +52,9 @@
-    group   => $group,
-    mode    => '0644',
+  case $::osfamily {
+    'Debian': { include ::marathon::config::sysvinit }
```
We need to contain or anchor these classes to be able to rely on their ordering relative to the other install steps.
Something like this should do:

``` puppet
'Debian': {
  include marathon::config::sysvinit
  contain 'marathon::config::sysvinit'
}
'Redhat': {
  include marathon::config:: systemd
  contain 'marathon::config:: systemd'
}
```

It's a pain.. but this is puppet :(

Also, just nitpicking.. but I'd prefer we stay consistent and don't use a leading `::` in front of class names. Maybe we should be, but in that case I'd rather make another PR and do it everywhere.

JayH5 commented on manifests/config.pp:
```diff
@@ -78,11 +52,9 @@
+    default: { fail("Unsupported OS family: ${::osfamily}") }
   }
-}
+}
```
Some of these files should have newlines at the end.

JayH5 commented on manifests/config/systemd.pp:
```diff
@@ -0,0 +1,69 @@
+    notify  => Exec['systemctl-daemon-reload_marathon'],
+  }
+
+  exec { 'systemctl-daemon-reload_marathon':
```
This seems like a thing that Puppet should handle? Or is there some standard module for handling systemd? I don't like the idea of writing `exec` commands to refresh services.

JayH5 commented on templates/sysconfig.erb:
```diff
@@ -0,0 +1,15 @@
+<% end if @env_var -%>
+
+<% @all_options.each do |k,v| -%>
+MARATHON_<%= k.upcase %>=<%= v %>
```
Is there any reason why you want to have all the config options here? I'd rather we provided the choice to the user of the module.

JayH5 commented on manifests/config/sysvinit.pp:
```diff
@@ -0,0 +1,56 @@
+  file { $conf_dir:
+    ensure => directory,
+    owner  => $owner,
+    group  => $group,
```
This is going to need to require `File[$marathon::config::conf_dir_base]`

JayH5 commented on spec/classes/config_spec.rb:
```diff
@@ -1,5 +1,155 @@
 require 'spec_helper'
```
:cake: :fireworks: :tada:

JayH5 commented:
Thanks again for the pull request.

The major thing I'd like to see is to still be able to configure Marathon using files in `/etc/marathon/conf` if that is what the user would like to do. I know it is simpler to define all the config as a bunch of variables in one file that is sourced... but the vast majority of services on Linux use configuration files in a `/etc/<service>` directory.. and my personal preference is to do that. Also, I'd prefer not to have a different way of configuring Marathon whether it's running on sysvinit or systemd - the init system isn't relevant to that.

felixb commented:
Thanks for your input. I'll work on that on Monday.

felixb commented:
I reverted the changes for handling config differently when running systemd, added some tests and modified the readme again.
I'll squash the commits after getting your ok for the changes.

marathon is configured in /etc/marathon for both init systems. Only things stored in `/etc/default/marathon` are configured in `/etc/sysconfig/marathon` when running systemd.

Regarding the systemd reload:
We need to reload systemd after adding/changing unit files. There are a few systemd modules available in the puppet forge. I can't use them right now because of some internal namespace clash..
I'll send a new PR as soon we got the things untangled here.

JayH5 commented on templates/sysconfig.erb:
```diff
@@ -0,0 +1,11 @@
+# This file is managed by Puppet
+
+<%- if !@java_home.nil? && !@java_home.empty? -%>
+JAVA_HOME=<%= @java_home %>
```
Please wrap the values for `JAVA_HOME` and `JAVA_OPTS` in `"`s.

felixb replied:
Same as above. This is valid syntax here.

JayH5 commented on templates/sysconfig.erb:
```diff
@@ -0,0 +1,11 @@
+JAVA_OPTS=<%= @java_opts %> <%= @java_extra_opts %>
+
+<% @env_var.sort.each do |key,val| -%>
+<%= key %>=<%= val %>
```
Also the values here.

felixb replied:
Same as above. This is valid syntax here.

JayH5 commented on spec/classes/config_spec.rb:
```diff
@@ -1,5 +1,140 @@
+
+            it 'configures systemd' do
+              should contain_file('/etc/sysconfig/marathon')
+                         .with_content(/^JAVA_OPTS=-Xmx1024m -Dlogback.configurationFile=file:\/etc\/marathon\/logback.xml$/)
```
If there are no quotations around `JAVA_OPTS` then it is really assigning `JAVA_OPTS=-Xmx1024m` and then running the rest as a command `-Dlogback....`

felixb replied:
That's actually valid systemd unit file syntax. Adding `"` would result in having `"-Xmx..."` as value.

JayH5 commented:
Regarding my comments above.. I may actually be wrong.. I'm assuming `/etc/sysconfig/marathon` is sourced/is plain shell and not some kind of config file.

At some point I should probably learn this systemd stuff...

felixb commented:
Systemd is not some bash magic. Unit files are not shell scripts sourced in like in sysvinit.

JayH5 commented:
OK, my mistake, sorry.

Have you tested this latest revision on an actual machine and does it work? If so, I'm happy to merge.

felixb commented:
I'm about to test it. I come back to you with a squashed and tested PR.

JayH5 commented:
Thanks!

JayH5 commented:
@felixb is this tested and ready for a final review?

felixb commented:
I just rolled out [this PR + the PR for secrets](https://github.com/felixb/puppet-marathon/tree/merged_features) to our live mesos cluster powered by 32 marathon nodes.

The same code was running in develop for ~0.5d now. Everything works as expected.
